geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/RangeHistogram.scala

```diff
@@ -25,61 +29,102 @@ import scala.reflect.ClassTag
  * @tparam T a comparable type which must have a StatHelperFunctions type class
  */
 class RangeHistogram[T](val attribute: Int, initialBins: Int, initialEndpoints: (T, T))
-                       (implicit defaults: MinMax.MinMaxDefaults[T], ct: ClassTag[T]) extends Stat {
+                       (implicit val defaults: MinMax.MinMaxDefaults[T], ct: ClassTag[T]) extends Stat {
 
   override type S = RangeHistogram[T]
 
   private [stats] var bins: BinnedArray[T] = BinnedArray[T](initialBins, initialEndpoints)
-  lazy private val stringify = Stat.stringifier(ct.runtimeClass, json = true)
+  lazy val stringify = Stat.stringifier(ct.runtimeClass)
+  private lazy val jsonStringify = Stat.stringifier(ct.runtimeClass, json = true)
 
   def length: Int = bins.length
+  def directIndex(value: Long): Int = bins.directIndex(value)
   def indexOf(value: T): Int = bins.indexOf(value)
   def count(i: Int): Long = bins.counts(i)
-  def endpoints: (T, T) = bins.bounds
+  def min: T = bins.bounds._1
+  def max: T = bins.bounds._2
+  def bounds: (T, T) = bins.bounds
+  def bounds(i: Int): (T, T) = bins.bounds(i)
   def medianValue(i: Int): T = bins.medianValue(i)
 
+  /**
+    * Copies another histogram into this one. In comparison to +=, this method will preserve the
+    * current bounds and length.
+    *
+    * @param other other histogram
+    */
+  def addCountsFrom(other: RangeHistogram[T]): Unit = {
+    if (length == other.length && bounds == other.bounds) {
+      this += other
+    } else {
+      RangeHistogram.copyInto(bins, other.bins)
+    }
+  }
+
   override def observe(sf: SimpleFeature): Unit = {
     val value = sf.getAttribute(attribute)
     if (value != null) {
-      bins.add(value.asInstanceOf[T])
+      val i = bins.indexOf(value.asInstanceOf[T])
+      if (i == -1) {
+        bins = RangeHistogram.expandBins(value.asInstanceOf[T], bins)
+        bins.add(value.asInstanceOf[T])
+      } else {
+        bins.counts(i) += 1
+      }
     }
   }
 
-  override def +(other: RangeHistogram[T]): RangeHistogram[T] = {
-    if (length == other.length && endpoints == other.endpoints) {
-      // hists match - we can just copy counts in directly
-      val plus = new RangeHistogram(attribute, length, endpoints)
-      var i = 0
-      while (i < plus.bins.length) {
-        plus.bins.counts(i) += (bins.counts(i) + other.bins.counts(i))
-        i += 1
+  override def unobserve(sf: SimpleFeature): Unit = {
+    val value = sf.getAttribute(attribute)
+    if (value != null) {
+      val i = bins.indexOf(value.asInstanceOf[T])
+      if (i == -1) {
```

Comment:
For this to happen, wouldn't we have to unobserve a feature that contains a value we've never seen before?  Is that reasonable?  (I'm uncertain in precisely what context we would call ```unobserve```, clearly.)